### PR TITLE
[DNM] Display Fat & Protein on Apple Watch

### PR DESCRIPTION
**Pull request by bashjs**
Added `Nutrients` screen where a user can log `Carbs`, `Protein`, and `Fat` from Apple Watch.

### Notes
- The new screen is disabled by default
- Picker step can be toggled between `5` and `1` via double tap for more precise numbers.

|iPhone|Apple Watch
|-|-
|Screenshot 2023-07-05 at 15 00 57|Screenshot 2023-07-05 at 01 15 11

### CODE REVIEW

File: FreeAPS/Sources/Services/WatchManager/WatchManager.swift [hunk 1]
```diff
@@ -5,9 +5,16 @@ import WatchConnectivity
...
+enum Nutrient: Decodable {
+    case carb
+    case protein
+    case fat
+}
+
 final class BaseWatchManager: NSObject, WatchManager, Injectable {
     private let session: WCSession
     private var state = WatchState()
+    private var addCarbsStateModel = AddCarbs.StateModel()
```
Comment (bashjs) [inline FreeAPS/Sources/Services/WatchManager/WatchManager.swift:17]: As a new to Swift, this was the trickiest part how to access the `.add()` method which performs FPU calculations.

File: FreeAPSWatch WatchKit Extension/Views/NutrientsView.swift [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+                return .system(size: 8)
+            case .small:
+                return .system(size: 9)
+            case .large,
+                 .medium:
+                return .system(size: 14)
+            default:
+                return .system(size: 16)
+            }
+        }
```
Comment (bashjs) [inline FreeAPSWatch WatchKit Extension/Views/NutrientsView.swift:132]: Numbers below 100 will use `fontForSizeCategory` scale, and all above 99 will use `fontForSizeCategoryForHundreds` accordingly. Tweaked for different sizes: 38, 40, 41, 44, 45 starting from Apple Watch Series 3.

File: FreeAPSWatch WatchKit Extension/Views/NutrientsView.swift [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+                            .font(value < 99 ? fontForSizeCategory : fontForSizeCategoryForHundreds)
+                            .foregroundColor(accentColor)
+                    }
+                } label: {
+                    var labelFrameTranslationX: CGFloat {
+                        let offset = geometry.size.width + Constants.pickerHSpacing
+                        if labelAlignment == .leading { return offset }
+                        if labelAlignment == .trailing { return -offset }
+                        return 0
+                    }
```
Comment (bashjs) [inline FreeAPSWatch WatchKit Extension/Views/NutrientsView.swift:150]: In order to give more space to each label. By default the label is limited by the width of its `Picker`.

Screenshot 2023-07-02 at 21 44 25

File: FreeAPSWatch WatchKit Extension/Views/NutrientsView.swift [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+            }
+        }
+
+        private func handlePickerDoubleTapGesture() {
+            step = step == Constants.defaultPickerStep ? 1 : Constants.defaultPickerStep
+            if step == Constants.defaultPickerStep {
+                selection = Int(Double(selection / Constants.defaultPickerStep).rounded(.down)) * Constants
+                    .defaultPickerStep
+            }
+        }
```
Comment (bashjs) [inline FreeAPSWatch WatchKit Extension/Views/NutrientsView.swift:175]: Double tap changes Picker step from 5 to 1 and vice versa. When the `step` is 5 we round the Picker value to the lower closest one divisible by 5.
An example: double tapping on a Picker when its value is `77` will round it to `75`

File: FreeAPS/Sources/Services/WatchManager/WatchManager.swift [hunk 5]
```diff
@@ -288,6 +296,30 @@ extension BaseWatchManager: WCSessionDelegate {
             }
         }
 
+        if let nutrientsData = message["addNutrients"] as? Data {
+            if let nutrients = try? JSONDecoder().decode([Nutrient: Int].self, from: nutrientsData) {
+                if nutrients.values.contains(where: { $0 > 0 }) {
+                    addCarbsStateModel.carbs = Decimal(nutrients[.carb] ?? 0)
+                    addCarbsStateModel.protein = Decimal(nutrients[.protein] ?? 0)
+                    addCarbsStateModel.fat = Decimal(nutrients[.fat] ?? 0)
+                    addCarbsStateModel.add()
```
Comment (bashjs) [inline FreeAPS/Sources/Services/WatchManager/WatchManager.swift:305]: 🚨 The app is crashing on this step

Comment (bashjs) [reply FreeAPS/Sources/Services/WatchManager/WatchManager.swift:305]: Screenshot 2023-07-05 at 16 36 21

**Comment by Jon-b-m** [comment]
There already was a WCSession for sending carbs to phone. Why not make another one for fat and protein also?

File: FreeAPS/Sources/Services/WatchManager/WatchManager.swift (hunk 5 above)
Comment (bashjs) [reply FreeAPS/Sources/Services/WatchManager/WatchManager.swift]: Fixed

File: FreeAPSWatch WatchKit Extension/Views/NutrientsView.swift (hunk 2 above)
Comment (bashjs) [reply FreeAPSWatch WatchKit Extension/Views/NutrientsView.swift]: Removed.

File: FreeAPS/Sources/Services/WatchManager/WatchManager.swift (hunk 1 above)
Comment (bashjs) [reply FreeAPS/Sources/Services/WatchManager/WatchManager.swift]: Removed

**Comment by bashjs** [comment]
> There already was a WCSession for sending carbs to phone. Why not make another one for fat and protein also? I think that would be the easiest way forward.

The handlers `watchStateModel.addCarbs` and the `WatchManager`'s `message["carbs"]` session listener have their own logic after submitting and receiving carbs. In order to keep them unchanges I just created a separate flow.
Otherwise, we couldn't just send two messages because of `apsManager.determineBasal()` at the end of each. Same for error handling when encoding / decoding the message.

File: FreeAPS/Sources/Services/WatchManager/WatchManager.swift [hunk 6]
```diff
@@ -288,6 +297,51 @@ extension BaseWatchManager: WCSessionDelegate {
...
+                    // Store the real carbs
+                    carbsStorage.storeCarbs([
+                        CarbsEntry(
+                            id: UUID().uuidString,
+                            createdAt: Date(),
+                            carbs: carbs,
+                            enteredBy: CarbsEntry.manual,
+                            isFPU: false, fpuID: nil
+                        )
+                    ])
```
Comment (bashjs) [inline FreeAPS/Sources/Services/WatchManager/WatchManager.swift:327]: It works the same way as [this method](https://github.com/Artificial-Pancreas/iAPS/blob/dev/FreeAPS/Sources/Modules/AddCarbs/AddCarbsStateModel.swift#L93-L108).

File: FreeAPS/Sources/Services/WatchManager/WatchManager.swift [hunk 7]
```diff
@@ -19,6 +25,8 @@ final class BaseWatchManager: NSObject, WatchManager, Injectable {
     @Injected() private var tempTargetsStorage: TempTargetsStorage!
     @Injected() private var garmin: GarminManager!
 
+    @Published var date = Date()
```
Comment (bashjs) [inline FreeAPS/Sources/Services/WatchManager/WatchManager.swift:28]: Not sure if `@Published` is required here. We could probably just inline the `Date()` where we use it.

File: FreeAPS/Sources/Modules/AddCarbs/AddCarbsStateModel.swift [hunk 8]
```diff
@@ -34,66 +34,11 @@ extension AddCarbs {
...
-                    let eachCarbEntry = CarbsEntry(
-                        id: UUID().uuidString, createdAt: useDate, carbs: equivalent, enteredBy: CarbsEntry.manual, isFPU: true,
-                        fpuID: fpuID
-                    )
-                    futureCarbArray.append(eachCarbEntry)
-                    numberOfEquivalents -= 1
-                }
-                // Save the array
-                if carbEquivalents > 0 {
+                let futureCarbArray = calculateFPU(settings: settings, protein: protein, fat: fat)
```
Comment (bashjs) [inline FreeAPS/Sources/Modules/AddCarbs/AddCarbsStateModel.swift:37]: Made `calculateFPU` as a simple helper function.
There is probably a bit more Swift-idiomatic way of creating such reusable pieces. Like a `class` or something.

**Comment by mountrcg** [comment]
using it since 2 weeks, works fine - great job

**Comment by bashjs** [comment]
Closing in favor of the current implementation which is in the `main` branch already.

**Pull request closed without merging**